packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/CollectionResultDefinition.cs

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.TypeSpec.Generator.ClientModel.Snippets;
+using Microsoft.TypeSpec.Generator.Input;
+using Microsoft.TypeSpec.Generator.Primitives;
+using Microsoft.TypeSpec.Generator.Providers;
+using Microsoft.TypeSpec.Generator.Snippets;
+using Microsoft.TypeSpec.Generator.Statements;
+using static Microsoft.TypeSpec.Generator.Snippets.Snippet;
+
+namespace Microsoft.TypeSpec.Generator.ClientModel.Providers
+{
+    internal class CollectionResultDefinition : TypeProvider
+    {
+        private readonly InputOperation _operation;
+        private readonly bool _isAsync;
+        private readonly ClientProvider _client;
+        private FieldProvider _clientField;
+        private readonly FieldProvider _optionsField;
+        private readonly CSharpType _responseType;
+        private readonly CSharpType? _modelType;
+        private readonly string? _nextLinkPropertyName;
+        private readonly FieldProvider _initialUri;
+        private readonly InputResponseLocation? _nextLinkLocation;
+
+        private static ParameterProvider PageParameter =
+            new("page", FormattableStringHelpers.Empty, new CSharpType(typeof(ClientResult)));
+
+        private readonly string _itemsPropertyName;
+
+        public CollectionResultDefinition(ClientProvider client, InputOperation operation, CSharpType? modelType, bool isAsync)
+        {
+            _client = client;
+            _clientField = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                _client.Type,
+                "_client",
+                this);
+            _initialUri = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                new CSharpType(typeof(Uri)),
+                "_initialUri",
+                this);
+            _optionsField = new FieldProvider(
+                FieldModifiers.Private | FieldModifiers.ReadOnly,
+                new CSharpType(typeof(RequestOptions)),
+                "_options",
+                this);
+
+            _modelType = modelType;
+            _operation = operation;
+            _isAsync = isAsync;
+
+            var response = _operation.Responses.FirstOrDefault(r => !r.IsErrorResponse);
+            var model = ScmCodeModelPlugin.Instance.TypeFactory.CreateModel((InputModelType)response!.BodyType!)!;
```

Comment:
so we have a `model` here, and there is an input argument `modelType`, what is the relationship between these two?
If they are the same thing (`model` is the typeprovider of `modelType`), do we still need the input `modelType`?
If they are different thing, their names here are confusing, we might need to change them.